doc/source/architecture.rst

```diff
@@ -0,0 +1,216 @@
+geometry, material properties, mesh, and connections to properly input them to
+the solver. As discussed earlier, the command pattern does not allow this kind
+of access. Due to this fact, the Mechanical API exposes its data model directly
+to the user. This is how a hypothetical command-based API would look for
+renaming an object.
+
+.. code:: python
+
+    RenameCommand(id=100, name="New name")
+
+Instead, an API based on a data model, like mechanical's API, looks like this:
+
+.. code:: python
+
+    obj = GetObject(id=100)
+    print(obj.Name)
+    obj.Name = "New name"
+
+Notice that you could print the name by Reading a property of the object. A
+command-based API can not provide the same experience.
+
+
+PyMechanical remote interface
+-----------------------------
+
+One way to interact with the Mechanical API from Python is as a remote session.
+You can run Mechanical as a server and send a *Remote Procedure Call* (RPC).
+The server handles the RPC and returns the response to the client. Currently,
+sending commands as a string and getting the result as a string is the only
+option.
+
+PyMechanical embedded instance
+--------------------------------
+
+The embedded instance used by PyMechanical embeds an entire instance of the
+Mechanical app in-memory inside of a Python program. There is no additional
+running process associated with it. Mechanical's data model is directly
+available within Python, which means that the fully CRUD data model of the
+mechanical API can be used.
+
+
+
+Distributed systems
+===================
+
+This section contains a very basic explanation of *distributed systems*. It is
+not meant be exhaustive and rigorous, but it instead introduces just the topics
+necessary to understand the choices made by the designers of PyMechanical.
+
+A distributed system is a software system that uses a network to distribute
+software across physical machines. With a distributed system, the individual
+pieces of that system do not share an address space and therefore cannot call
+functions of each other directly. Instead, they communicate with each other by
+sending messages in packets.
+
+Distributed systems have unique characteristics when compared to classical
+software systems that share an address space. For instance, in a distributed
+system, any call can fail because of a problem with the network, and the caller
+can not always know whether a call has failed. For this reason, features of
+interest to distributed system designers, such as fault tolerance and
+idempotency, are not emphasized by classical software systems.
+
+Examples include email, multiplayer games, web apps, and high-performance
+computing, among other things. Historically, there are three major categories
+of remote APIs: message passing, resource based, and remote method invocation.
+
+Message passing
+---------------
+
+Message passing is the original model of remote APIs, and in fact the other two
```

Comment:
what other two categories? Resource-based and remote method invocation APIs?

Reply:
Kerry's confusion here is why I relates back to here: categories of remote APIs: message-passing, resource-based, and remote method invocation.

It's basically why I suggested that the following section be named as follows:
Message-passing APIs
Resource-based APIs
Remote Method Invocation APIs

If we define Remote Method Invocation as RMI ealier, we could then RMI APIs to shorten the third heading.

Reply:
@PipKat @klmcadams I can't call them APIs because they aren't... but I will do my best to rework/clarify it.

Reply:
@PipKat @klmcadams I overhauled this section. I think its easier to understand, as well as being more acurrate. Let me know what you think.